parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt

```diff
@@ -0,0 +1,158 @@
+        val secondEvent = tracker.events[1]
+        assertThat(secondEvent).isCorrectEvent(
+            // Ideally: totalTime should be equal to DEFAULT_INTERVAL_MILLIS * 2
+            withTotalTime = { isCloseTo(DEFAULT_INTERVAL_MILLIS * 2, withinPercentage(10)) },
+            // Ideally: timestamp should be equal to `now` at the time of recording the event
+            withTimestamp = { isCloseTo(secondTimestamp, within(20L)) }
+        )
+        val thirdEvent = tracker.events[2]
+        assertThat(thirdEvent).isCorrectEvent(
+            // Ideally: totalTime should be equal to DEFAULT_INTERVAL_MILLIS * 3
+            withTotalTime = { isCloseTo(DEFAULT_INTERVAL_MILLIS * 3, withinPercentage(10)) },
+            // Ideally: timestamp should be equal to `now` at the time of recording the event
+            withTimestamp = { isCloseTo(thirdTimestamp, within(20L)) }
+        )
+    }
+
+    private fun MapAssert<String, Any>.isCorrectEvent(
+        withTotalTime: AbstractLongAssert<*>.() -> AbstractLongAssert<*>,
+        withTimestamp: AbstractLongAssert<*>.() -> AbstractLongAssert<*>,
+    ): MapAssert<String, Any> {
+        return containsEntry("action", "heartbeat")
+            // Incremental will be always 0 because the interval is lower than 1s
+            .containsEntry("inc", 0L)
+            .hasEntrySatisfying("tt") { totalTime ->
+                totalTime as Long
+                assertThat(totalTime).withTotalTime()
+            }
+            .hasEntrySatisfying("data") { data ->
+                data as Map<String, Any>
+                assertThat(data).hasEntrySatisfying("ts") { timestamp ->
+                    timestamp as Long
+                    assertThat(timestamp).withTimestamp()
+                }.containsAllEntriesOf(testData.minus("ts"))
+            }
+    }
+
+    class FakeTracker : ParselyTracker(
+        "",
+        0,
+        ApplicationProvider.getApplicationContext()
+    ) {
+        val events = mutableListOf<Event>()
+
+        override fun enqueueEvent(event: Event) {
+            events += event
+        }
+    }
+
+    class FakeIntervalCalculator : UpdateEngagementIntervalCalculator() {
+        override fun updateLatestInterval(startTime: Calendar): Long {
+            return DEFAULT_INTERVAL_MILLIS
+        }
+    }
+
+    private val now: Long
+        get() = Calendar.getInstance(TimeZone.getTimeZone("UTC")).timeInMillis
+
+    private fun sleep(millis: Long) = Thread.sleep(millis + THREAD_SLEEPING_THRESHOLD)
+
+    private companion object {
+        const val DEFAULT_INTERVAL_MILLIS = 100L
+        // Additional time to wait to ensure that the timer has fired
+        const val THREAD_SLEEPING_THRESHOLD = 50L
+        val testData = mutableMapOf<String, Any>(
+            "os" to "android",
+            "parsely_site_uuid" to "e8857cbe-5ace-44f4-a85e-7e7475f675c5",
+            "os_version" to "34",
+            "manufacturer" to "Google",
+            "ts" to 123L
+        )
+    }
+}
```

Comment:
Minor (🔍): Consider adding an empty line at the end of this test.

FYI: You can enable the `Ensure every saved file ends with a line break` option within your IDE.

PS: You will find that within `Editor` -> `General` -> `On Save` section at the bottom.